src/p11_wrap.cpp

```diff
@@ -0,0 +1,276 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2022 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+/*
+ * This file is the only place where we should see any
+ * methods from LibP11. Any other code should use
+ * the more high-level methods declared and define in
+ * p11_wrap.cpp and p11_wrap.h.
+ *
+ */
+
+#include <boost/algorithm/hex.hpp>
+#include <cassert>
+#include <cstddef> /* this has to come before cppc (bug in boost) */
+#include <exception>
+
+#include <cppc/checkcall.hpp>
+#include "mococrw/openssl_wrap.h"
+#include "mococrw/p11_wrap.h"
+
+namespace mococrw
+{
+namespace p11
+{
+// Initialise error strings of LibP11 wrapper:
+const std::string P11Exception::nullCtxExceptionString = "Found NULL PKCS11 Context";
+const std::string P11Exception::nullSlotListExceptionString = "Found NULL Slot List";
+const std::string P11Exception::nullTokenExceptionString = "Found NULL Token";
+const std::string P11Exception::mismatchLoginExceptionString = "Login/Logout Mismatch Detected";
+
+/**
+ * This struct is used by CWrap to
+ * determine how to react to an error
+ * when performing a LibP11 call.
+ */
+struct P11ExceptionErrorPolicy
+{
+    template <class Rv>
+    static void handleError(const Rv &);
+};
+
+/**
+ * Throw a P11 exception upon error.
+ *
+ * This method fetches the error string and throws an exception
+ * with the corresponding message.
+ */
+template <class Rv>
+void P11ExceptionErrorPolicy::handleError(const Rv & /* unused */)
+{
+    throw P11Exception();
+}
+
+// For checking return values:
+using P11CallPtr =
+        ::cppc::CallCheckContext<::cppc::IsNotNullptrReturnCheckPolicy, P11ExceptionErrorPolicy>;
+using P11IsNonNegative =
+        ::cppc::CallCheckContext<::cppc::IsNotNegativeReturnCheckPolicy, P11ExceptionErrorPolicy>;
+
+/***********************************************************************************
+ * Implementation of P11_SlotInfo
+ */
+
+P11_SlotInfo::P11_SlotInfo(P11_PKCS11_CTX_Ptr ctx) : _ctx(ctx), _slotList(nullptr), _numSlots(0)
```

Comment:
Can this be somehow a part of P11_Session class? I'm not sure why user of this  API needs to think about constructing 2 different objects before doing anything with HSM.

Reply:
It was something which I did think of in my initial design, but I think this current approach is better. The reason is that enumeration is done once for slot selection, whereas sessions are created for every operation. In other words, we do not want to perform slot enumeration every time we perform an operation on the HSM.

Reply:
Consider using isEnumerated flag in constructor then to simplify design.

Reply:
I'm not sure if this is worth the effort, and I'm not sure I like the prospect of having this class's constructor stateful based on a flag.

This API is an internal one and is not a concern for the user.